tests/src/tests/test_utils.go

```diff
@@ -0,0 +1,324 @@
+func init() {
+	if _, err := os.Stat(tmpDir); !os.IsNotExist(err) {
+		os.RemoveAll(tmpDir)
+	}
+	os.Mkdir(tmpDir, 0755)
+
+	f, err := os.OpenFile(filepath.Join(tmpDir, logFileName), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		fmt.Printf("Error creating Log file : %v", err)
+	} else {
+		if logToFileOnly {
+			testLogger = log.New(f, "", log.LstdFlags|log.Lmicroseconds)
+		} else {
+			writer := io.MultiWriter(f, os.Stdout)
+			testLogger = log.New(writer, "", log.LstdFlags|log.Lmicroseconds)
+		}
+		testLogger.Println("Test Starting:")
+	}
+}
+
+func GetTempDir() string {
+	_, fn, _, ok := runtime.Caller(1)
+	if !ok {
+		return tmpDir
+	}
+	testFile := filepath.Base(fn)
+	testFileExtension := filepath.Ext(testFile)
+	subdir := testFile[0 : len(testFile)-len(testFileExtension)]
+	return CreateTempDir(subdir)
+}
+
+func CreateTempDir(subdir string) string {
+	tempDir := tmpDir + subdir + "/"
+	if _, err := os.Stat(tempDir); os.IsNotExist(err) {
+		os.Mkdir(tempDir, 0755)
+	}
+	return tempDir
+}
+
+func GetDevFileName() string {
+	pc, fn, _, ok := runtime.Caller(1)
+	if !ok {
+		return tmpDir + "DefaultDevfile"
+	}
+
+	testFile := filepath.Base(fn)
+	testFileExtension := filepath.Ext(testFile)
+	subdir := testFile[0 : len(testFile)-len(testFileExtension)]
+	destDir := CreateTempDir(subdir)
+	callerName := runtime.FuncForPC(pc).Name()
+
+	pos1 := strings.LastIndex(callerName, "/tests.") + len("/tests.")
+	devfileName := destDir + callerName[pos1:len(callerName)] + ".yaml"
+
+	LogMessage(fmt.Sprintf("GetDevFileName : %s", devfileName))
+
+	return devfileName
+}
+
+func AddSuffixToFileName(fileName string, suffix string) string {
+	pos1 := strings.LastIndex(fileName, ".yaml")
+	newFileName := fileName[0:pos1] + suffix + ".yaml"
+	LogMessage(fmt.Sprintf("Add suffix %s to fileName %s : %s", suffix, fileName, newFileName))
+	return newFileName
+}
+
+func LogMessage(message string) string {
+	testLogger.Println(message)
+	return message
+}
+
+type TestDevfile struct {
+	SchemaDevFile   schema.Devfile
+	FileName        string
+	ParsedSchemaObj parser.DevfileObj
+	SchemaParsed    bool
+}
+
+var StringCount int = 0
+
+var RndSeed int64 = time.Now().UnixNano()
+
+func GetRandomUniqueString(n int, lower bool) string {
+	StringCount++
+	return fmt.Sprintf("%s%04d", GetRandomString(n, lower), StringCount)
+}
+
+func setRandSeed() {
+	RndSeed++
+	rand.Seed(RndSeed)
+}
+
+const schemaBytes = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
+
+func GetRandomString(n int, lower bool) string {
+	setRandSeed()
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = schemaBytes[rand.Intn(len(schemaBytes)-1)]
+	}
+	randomString := string(b)
+	if lower {
+		randomString = strings.ToLower(randomString)
+	}
+	return randomString
+}
+
+func GetGroupKinds() []schema.CommandGroupKind {
+	setRandSeed()
+	return []schema.CommandGroupKind{schema.BuildCommandGroupKind, schema.RunCommandGroupKind, schema.TestCommandGroupKind, schema.DebugCommandGroupKind}
```

Comment:
this is a constant slice, no need to setRandSeed. and no need a separate function for the group kinds
suggest to have the slice defined at top of the file, as you've done for protocol and exposure in endpoints

Reply:
Yes, I created endpoint later than components - maybe my go skills were increasing :-)

Reply:
Done